src/visitors/globals.js

```diff
@@ -7,7 +7,8 @@ const VARS = {
     asset.addDependency('process');
     return 'var process = require("process");';
   },
-  global: () => 'var global = (1,eval)("this");',
+  global: () =>
+    'var global = typeof global === "object" ? global : typeof window === "object" ? window : typeof self === "object" ? self : {};',
```

Comment:
I don't think this is necessarily very reliable. What if someone declares `window` or `self` in the local scope of their module?

Reply:
That's a good point. Replaced it with a more reliable (and succinct!) solution: https://github.com/parcel-bundler/parcel/pull/1133/commits/1fe753ed5d276eeaf8619100d6bd22aa15eb27fe

Reply:
Yeah but won't `Function` constructor have the same problem as `eval` with CSP?

Reply:
You're, right, my bad.
I don't know why I assumed that another variant of arbitrary code execution would fly. 🙃  Should have read [W3C Working Draft for CSP](https://www.w3.org/TR/CSP/#directive-script-src)…

About the case of `window` and `self` in a local scope – you're talking about a situation where there's a module that either:
1. uses `global` object, redefines `window`, and is used in a browser
2. uses `global` object, redefines `self`, and is used in a Web Worker

Do we have to choose between supporting those two cases and supporting CSP?

Reply:
Yeah... we may be able to get it working by moving that code to get the global object you originally added outside of each module and into the prelude where we can guarantee that nothing will mess with those variables. In there, the global object should be `this` (unless strict mode is enabled). Then we can attach it to `module` or something, and here assign `var global = module.global`, or maybe pass it into the module function. What do you think?

See this article for more info: https://www.contentful.com/blog/2017/01/17/the-global-object-in-javascript/